Tasks 1,2,3 Completed

graceestewart opened the pull request:
Tasks 1, 2, and 3 (listed below for convenience) from the starter project have been completed. Each task has its own commit. 

Task 1:  Implement the ‘Create’ Data Access Object (DAO) method for Inventory objects with unit testing
Task 2:  Implement the ‘Create’ Rest Controller method for Inventory objects with unit testing
Task 3:  Implement the ‘Delete’ Data Access Object (DAO) method for Inventory objects with unit testing

graceestewart committed:
completed task1 in starter project

graceestewart committed:
Completed task2 in the starter project

graceestewart committed:
Completed task3 in the starter project

jinlin0 commented on starter/src/main/java/com/starter/fullstack/dao/InventoryDAO.java:
```diff
@@ -82,7 +83,11 @@ public Optional<Inventory> update(String id, Inventory inventory) {
-    // TODO
-    return Optional.empty();
+    Inventory deleted = this.mongoTemplate.findById(id, Inventory.class);
+    if(deleted == null){return null;}
```
Since the return type is an Optional, you'll want to get familiar with how to return Optional and what they're used for. I've linked a page we wrote up to give a short introduction. 

https://sofiac.atlassian.net/wiki/spaces/TS/pages/176619685/Quality+Checkstyle+Assurance+Guide#Optional

graceestewart replied:
For the return value, it won't let me return a non-optional, but if the value is null, I cannot using Optional.of(), as that throws an error. This being said, because findandremove returns an Inventory item (non-optional) or null, I also can't set Option.of() to the result of findandremove, and I cannot set an optional variable to be equal to the result of findandremove, so I still have a quick check if the result of findandremove is null. I will keep looking however. Also, I know that Optional<Inventory> can be null or an Inventory object, but I get an error saying that I need to return an optional type, not inventory, which is why I am doing the check still.

jinlin0 replied:
There is a method within Optional that should allow you to do what you're looking for 

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

jinlin0 commented on starter/src/main/java/com/starter/fullstack/dao/InventoryDAO.java:
```diff
@@ -82,7 +83,11 @@ public Optional<Inventory> update(String id, Inventory inventory) {
   public Optional<Inventory> delete(String id) {
-    // TODO
-    return Optional.empty();
+    Inventory deleted = this.mongoTemplate.findById(id, Inventory.class);
```
This does work, but I'd challenge you to look through mongoTemplate's documentation to see if you can find a way to find and delete with a single database call.

You always want to try to make the least amount of database calls to maximize performance

graceestewart replied:
I rewrote this to now use findandremove, so I'm only making one database call

jinlin0 commented on starter/src/test/java/com/starter/fullstack/dao/InventoryDAOTest.java:
```diff
@@ -52,4 +53,117 @@ public void findAll() {
+
+  // make sure that it can insert more than one object
+  @Test
+  public void create2() {
```
Great pass at writing these tests!

You don't necessarily need to write a test to call the function twice unless the behavior somehow changes. For example, if you have an if statement in the function, you'll want to test what happens if the if statement is true and if the statement is false. 

But since you don't have any branching logic in the function, just 1 test is fine

graceestewart replied:
combined tests (multiple asserts in one test now) so there is only create1

jinlin0 commented on starter/src/test/java/com/starter/fullstack/rest/InventoryControllerTest.java:
```diff
@@ -0,0 +1,112 @@
+
+    // check to make sure that it sets the ID to null
+    @Test
+    public void create2() throws Throwable {
```
For this test, it's a great assert

However since the input to the function doesn't change, it's technically the first test with a different assert. In cases where you want to assert on multiple things, but the input is the same, feel free to combine the 2 tests together and add additional asserts

graceestewart replied:
changed to checking this in create1

graceestewart committed:
Completed task 4, changed the delete funciton in inventoryDAO, changed tests per previous PR comments

graceestewart commented:
I made a new commit where I completed task 4, I combined all create tests into one test, I rewrote the delete function using findandremove, and I rewrote the delete tests (there are only 2 now).
